agent/agent_test.go

```diff
@@ -199,72 +199,125 @@ func TestAgent_Stats_Magic(t *testing.T) {
-		sshClient, err := conn.SSHClient(ctx)
-		require.NoError(t, err)
+	// The echo server prints its port as the first line.
+	sc := bufio.NewScanner(stdout)
+	sc.Scan()
+	remotePort := sc.Text()
 
-		tunneledConn, err := sshClient.Dial("tcp", fmt.Sprintf("127.0.0.1:%s", remotePort))
-		require.NoError(t, err)
-		t.Cleanup(func() {
-			// always close on failure of test
-			_ = conn.Close()
-			_ = tunneledConn.Close()
-		})
+	return remotePort, echoServerCmd
+}
 
-		require.Eventuallyf(t, func() bool {
-			s, ok := <-stats
-			t.Logf("got stats with conn open: ok=%t, ConnectionCount=%d, SessionCountJetBrains=%d",
-				ok, s.ConnectionCount, s.SessionCountJetBrains)
-			return ok && s.ConnectionCount > 0 &&
-				s.SessionCountJetBrains == 1
-		}, testutil.WaitLong, testutil.IntervalFast,
-			"never saw stats with conn open",
-		)
+func verifyJetBrainsTracking(t *testing.T, network, address string) {
+	ctx := testutil.Context(t, testutil.WaitLong)
 
-		// Kill the server and connection after checking for the echo.
-		requireEcho(t, tunneledConn)
-		_ = echoServerCmd.Process.Kill()
-		_ = tunneledConn.Close()
+	//nolint:dogsled
+	conn, _, stats, _, _ := setupAgent(t, agentsdk.Manifest{}, 0)
+	sshClient, err := conn.SSHClient(ctx)
+	require.NoError(t, err)
 
-		require.Eventuallyf(t, func() bool {
-			s, ok := <-stats
-			t.Logf("got stats after disconnect %t, %d",
-				ok, s.SessionCountJetBrains)
-			return ok &&
-				s.SessionCountJetBrains == 0
-		}, testutil.WaitLong, testutil.IntervalFast,
-			"never saw stats after conn closes",
-		)
+	// On Linux it is permissible to listen on the same port on different
+	// addresses, and when this happens `localhost` can route to the ipv6 loopback
+	// address instead of the ipv4 address.  Since we make requests to coderd with
+	// `localhost` this behavior can cause those requests to go to the ipv6 echo
+	// server instead of coderd when there is a port collision.  Listening on ipv4
+	// here occupies the port, preventing coderd from using it and running into
+	// this issue.
+	port := "0"
+	if network == "tcp6" {
+		port, _ = spawnEchoServer(t, "tcp4", "0")
+	}
+
+	// JetBrains tracking works by looking at the process name listening on
+	// the forwarded port.  If the process's command line includes the magic
+	// string we are looking for, then we assume it is a JetBrains editor.
+	// So when we connect to the port we must ensure the process includes
+	// that magic string to fool the agent into thinking this is JetBrains.
+	// To do this we need to spawn an external process (in this case a
+	// simple echo server) so we can control the process name.  The -D here
+	// is just to mimic how Java options are set but is not necessary as the
+	// agent looks only for the magic string itself anywhere in the command.
+	gotPort, echoServerCmd := spawnEchoServer(t, network, port, "-D", agentssh.MagicProcessCmdlineJetBrains)
+	if port != "0" {
+		require.Equal(t, gotPort, port) // Sanity check.
```

Comment:
This creates 2 echo servers, which could have different lifetimes, possibly leading us back to the same flakes (albeit much much more rarely).

Also, you could bind to IPv4 above and then find the IPv6 port is in use and then flake on this line.

I think if you listen to `network="tcp"` and `address="[::1]:0"` then you get a socket that listens on v4 and v6 loopback on the same port. I think you should have the Echo server always do this, since we're testing the dialed connection, not the listener, right?

Reply:
Oh that is way better.  And yup good point about testing the connection, not the listener.

Reply:
Oh right one reason to have them separate is to make sure the code is not simply picking up the ipv4 port, so the test would no longer guarantee we actually are finding the ipv6 port.  Maybe we will need to change those`localhost` requests instead of double-binding here after all.